os/drivers/usbhost/uvc/usbhost_uvc_driver.c

```diff
@@ -0,0 +1,2377 @@
+ */
+
+#define DEV_FORMAT          "/dev/video%d"
+#define DEV_NAMELEN         12
+
+extern struct video_devops_s g_uvc_video_devops;
+
+static struct uvc_format_desc uvc_fmts[] = {
+	{"RGB565", UVC_GUID_FORMAT_RGBP, V4L2_PIX_FMT_RGB565},
+	{"MJPEG", UVC_GUID_FORMAT_MJPEG, V4L2_PIX_FMT_MJPEG},
+	{"YUV 4:2:2 (UYVY)", UVC_GUID_FORMAT_UYVY, V4L2_PIX_FMT_UYVY},
+	{"YUY2", UVC_GUID_FORMAT_YUY2, V4L2_PIX_FMT_YUY2}
+};
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static void uvc_stream_delete(struct uvc_streaming *stream);
+static int uvc_register_camera(FAR uvc_state_t *dev);
+
+/* Memory allocation services */
+
+static inline FAR struct usbhost_uvc_state_s *usbhost_uvc_allocclass(void);
+static inline void usbhost_uvc_freeclass(FAR struct usbhost_uvc_state_s *usbclass);
+
+/* Device name management */
+
+static int usbhost_uvc_allocdevno(FAR struct usbhost_uvc_state_s *priv);
+static void usbhost_uvc_freedevno(FAR struct usbhost_uvc_state_s *priv);
+static inline void usbhost_uvc_mkdevname(FAR struct usbhost_uvc_state_s *priv, FAR char *devname);
+
+/* Worker thread actions */
+static void usbhost_uvc_destroy(FAR void *arg);
+
+/* Helpers for usbhost_uvc_connect() */
+
+static int usbhost_uvc_cfgdesc(FAR struct usbhost_uvc_state_s *priv, FAR const uint8_t *configdesc, int desclen);
+static inline int usbhost_uvc_devinit(FAR struct usbhost_uvc_state_s *priv);
+
+/* Transfer descriptor memory management */
+static int usbhost_alloc_buffers(FAR struct usbhost_uvc_state_s *priv);
+static void usbhost_free_buffers(FAR struct usbhost_uvc_state_s *priv);
+
+/* struct usbhost_uvc_registry_s methods */
+
+static struct usbhost_class_s *usbhost_uvc_create(FAR struct usbhost_hubport_s *hport, FAR const struct usbhost_id_s *id);
+
+/* struct usbhost_class_s methods */
+
+static int usbhost_uvc_connect(FAR struct usbhost_class_s *usbclass, FAR const uint8_t *configdesc, int desclen);
+static int usbhost_uvc_disconnected(FAR struct usbhost_class_s *usbclass);
+
+/* Driver methods -- depend upon the type of TinyAra driver interface exported */
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* This structure provides the registry entry ID information that will  be
+ * used to associate the USB class driver to a connected USB device.
+ */
+
+static const const struct usbhost_id_s g_id = {
+	USB_CLASS_MISC,				/* base -- Must be one of the USB_CLASS_* definitions in usb.h */
+	0x02,						/* subclass -- depends on the device */
+	0x01,						/* proto -- depends on the device    */
```

Comment:
we'd better define subclass & proto for uvc host driver.

Reply:
Ideally these macros should be part of TizenRT USB host header files, but unfortunately those are missing. I have added UVC_DEVICE_SUBCLASS & UVC_DEVICE_PROTOCOL to uvc header files.